### PR TITLE
FileWalkers refactoring

### PR DESCRIPTION
**Pull request by ov7a**
### Context
This refactoring arose as a part of #3982. 

1. ReproducibleFileWalker can use the same abstractions as DefaultDirectoryWalker
2. All file walkers should use attributes-based fileVisitDetails for consistency
3. SingleIncludePatternFileTree should pass `stopFlag` properly to DirectoryFileTree

I considered refactoring SingleIncludePatternFileTree to use PathVisitor, but it seems like it's not worth it.

### CODE REVIEW

**Comment by ov7a** [comment]
@bot-gradle test this

**Comment by ov7a** [comment]
@bot-gradle test APT

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/SingleIncludePatternFileTree.java [hunk 1]
```diff
@@ -81,10 +82,10 @@ public void visitStructure(MinimalFileTreeStructureVisitor visitor, FileTreeInte
 
     @Override
     public void visit(FileVisitor visitor) {
-        doVisit(visitor, baseDir, new LinkedList<>(), 0, new AtomicBoolean());
+        doVisit(visitor, baseDir, new ArrayDeque<>(), 0, new AtomicBoolean());
     }
 
-    private void doVisit(FileVisitor visitor, File file, LinkedList<String> relativePath, int segmentIndex, AtomicBoolean stopFlag) {
+    private void doVisit(FileVisitor visitor, File file, Deque<String> pathSegments, int segmentIndex, AtomicBoolean stopFlag) {
```
Comment (lptr) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/SingleIncludePatternFileTree.java:88]: 🧐 I like the change to `ArrayDeque` but why rename the parameter? It _is_ holding a relative path, no?

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AttributeBasedFileVisitDetailsFactory {
```
Comment (lptr) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java:33]: 🧐 In my experience, classes with static methods like this are usually called `Utils` in our codebase; factories are passed around as instances (typically of `Factory`).

Comment (lptr) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java:33]: :x: This class and the methods could use some Javadocs.

We seem to have added quite a bit of code in this commit, without adding functionality. 🤔 I'm wondering if we need all of the methods, and if we need all of them to be public.

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/SingleIncludePatternFileTree.java (hunk 1 above)
Comment (ov7a) [reply subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/SingleIncludePatternFileTree.java:88]: It is confusing. I expect `relativePath` to have RelativePath type, not some kind of array

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java (hunk 2 above)
Comment (ov7a) [reply subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java:33]: My intention here is to convey that this class is used to create an instance, nothing more. `Utils` is too generic for this IMO

Comment (ov7a) [reply subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AttributeBasedFileVisitDetailsFactory.java:33]: These methods are just extracted from the existing code with some deduplication. 
Three of them are used. The fourth one is added for completeness, I see no harm in it, given that it's only two lines.

I agree with adding Javadoc, though. I'll do it later.

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Pull request merged**